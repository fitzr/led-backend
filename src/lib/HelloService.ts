import core = require('@aws-cdk/core')
import apigateway = require('@aws-cdk/aws-apigateway')
import lambda = require('@aws-cdk/aws-lambda')
import path = require('path')

export default class HelloService extends core.Construct {
  constructor(scope: core.Construct, id: string) {
    super(scope, id)

    const handler = new lambda.Function(this, 'hello-lambda', {
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'Hello.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../lambda'))
    })

    const api = new apigateway.RestApi(this, 'hello-api', {
      restApiName: 'Hello Service',
      description: 'Test Gateway'
    })

    const getHelloIntegration = new apigateway.LambdaIntegration(handler, {
      requestTemplates: { 'application/json': '{"statusCode":200}' }
    })

    api.root.addMethod('GET', getHelloIntegration)
  }
}
