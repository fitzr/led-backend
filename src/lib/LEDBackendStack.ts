import core = require('@aws-cdk/core')
import LEDBackendAPI from './LEDBackendApi'

export default class LEDBackendStack extends core.Stack {
  constructor(scope: core.Construct, id: string, props?: core.StackProps) {
    super(scope, id, props)

    new LEDBackendAPI(this, 'LEDBackendApi')
  }
}
