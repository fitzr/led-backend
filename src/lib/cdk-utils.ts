import * as cdk from '@aws-cdk/core'

export class CdkUtils {
  public static getEnv(scope: cdk.Construct): string {
    const env = scope.node.tryGetContext('env')
    if (!env) {
      throw new Error('Context "env" was not defined.')
    }
    return env
  }

  public static makeId(scope: cdk.Construct, prefix: string): string {
    return `${prefix}-${CdkUtils.getEnv(scope)}`
  }
}
