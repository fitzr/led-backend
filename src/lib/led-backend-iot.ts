import * as cdk from '@aws-cdk/core'
import * as iot from '@aws-cdk/aws-iot'
import * as iam from '@aws-cdk/aws-iam'
import { helper } from './stack-helper'

export class LedBackendIot extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string) {
    super(scope, id)
    this.createPolicy()
  }

  private createPolicy(): void {
    const account = cdk.Stack.of(this).account
    const policyDocument = new iam.PolicyDocument({
      statements: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['iot:Connect'],
          resources: [
            `arn:aws:iot:${helper.region}:${account}:client/\${iot:Connection.Thing.ThingName}`
          ]
        }),
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['iot:Publish', 'iot:Receive'],
          resources: [
            `arn:aws:iot:${helper.region}:${account}:topic/$aws/things/\${iot:Connection.Thing.ThingName}/shadow/*`
          ]
        }),
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['iot:Subscribe'],
          resources: [
            `arn:aws:iot:${helper.region}:${account}:topicfilter/$aws/things/\${iot:Connection.Thing.ThingName}/shadow/*`
          ]
        })
      ]
    })

    new iot.CfnPolicy(this, 'IotPolicy', {
      policyName: 'PubSubShadowPolicy',
      policyDocument
    })
  }
}
