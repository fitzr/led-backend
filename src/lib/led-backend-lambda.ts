import * as cdk from '@aws-cdk/core'
import * as lambda from '@aws-cdk/aws-lambda'
import * as iam from '@aws-cdk/aws-iam'
import { helper } from './stack-helper'

export class LedBackendLambda extends cdk.Construct {
  public readonly getStatusFunction: lambda.Function
  public readonly sendMessageFunction: lambda.Function

  constructor(scope: cdk.Construct, id: string) {
    super(scope, id)

    const environment = {
      region: helper.region,
      endpoint: helper.iotEndpoint
    }

    this.getStatusFunction = new lambda.Function(this, 'GetStatusFunction', {
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'get-status.handler',
      code: lambda.Code.fromAsset('src/lambda'),
      environment
    })

    this.sendMessageFunction = new lambda.Function(
      this,
      'SendMessageFunction',
      {
        runtime: lambda.Runtime.NODEJS_12_X,
        handler: 'send-message.handler',
        code: lambda.Code.fromAsset('src/lambda'),
        environment
      }
    )
    this.sendMessageFunction.addToRolePolicy(
      new iam.PolicyStatement({
        resources: [`arn:aws:iot:${helper.region}:*`],
        actions: ['iot:Publish']
      })
    )
  }
}
