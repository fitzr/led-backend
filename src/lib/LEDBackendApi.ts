import { Construct } from '@aws-cdk/core'
import { Code, Runtime, Function } from '@aws-cdk/aws-lambda'
import { RestApi, LambdaIntegration } from '@aws-cdk/aws-apigateway'
import { Resource } from '@aws-cdk/aws-apigateway/lib/resource'
import CDKUtils from './CDKUtils'
import path = require('path')

export default class LEDBackendApi extends Construct {
  constructor(scope: Construct, id: string) {
    super(scope, id)
    const apiVersion = scope.node.tryGetContext('apiVersion') || 'v1'
    const api = this.createApi(scope)
    const resource = api.root.addResource(apiVersion)
    this.addHelloApi(resource)
    this.addSendMessageApi(resource)
  }

  private createApi(scope: Construct): RestApi {
    const env = CDKUtils.getEnv(scope)
    const api = new RestApi(this, 'RestApi', {
      restApiName: `led-backend-api-${env}`,
      description: 'APIs for LED app.',
      deployOptions: {
        stageName: env
      }
    })
    const key = api.addApiKey('ApiKey')
    const plan = api.addUsagePlan('UsagePlan', {
      name: 'no-throttle',
      apiKey: key
    })
    plan.addApiStage({ stage: api.deploymentStage })
    return api
  }

  private addHelloApi(resource: Resource): void {
    const func = new Function(this, 'HelloFunction', {
      runtime: Runtime.NODEJS_12_X,
      handler: 'Hello.handler',
      code: Code.fromAsset(path.join(__dirname, '../lambda'))
    })
    const integration = new LambdaIntegration(func, {})
    resource.addResource('hello').addMethod('GET', integration, {
      apiKeyRequired: true
    })
  }

  private addSendMessageApi(resource: Resource): void {
    const func = new Function(this, 'SendMessageFunction', {
      runtime: Runtime.NODEJS_12_X,
      handler: 'SendMessage.handler',
      code: Code.fromAsset(path.join(__dirname, '../lambda'))
    })
    const integration = new LambdaIntegration(func, {})
    resource.addResource('send-message').addMethod('POST', integration, {
      apiKeyRequired: true
    })
  }
}
