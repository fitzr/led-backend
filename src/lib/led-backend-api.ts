import * as cdk from '@aws-cdk/core'
import * as lambda from '@aws-cdk/aws-lambda'
import * as apigw from '@aws-cdk/aws-apigateway'
import { helper } from './stack-helper'
import { LedBackendLambda } from './led-backend-lambda'

export class LedBackendApi extends cdk.Construct {
  static readonly API_VERSION = 'v1'

  constructor(scope: cdk.Construct, id: string, lambda: LedBackendLambda) {
    super(scope, id)
    const api = this.createApi()
    const resource = api.root
      .addResource(LedBackendApi.API_VERSION)
      .addResource('leds')
      .addResource('{thing_name}')
    this.addGetStatusMethod(resource, lambda.getStatusFunction)
    this.addSendMessageMethod(resource, lambda.sendMessageFunction)
  }

  private createApi(): apigw.RestApi {
    let props: apigw.RestApiProps = {
      restApiName: helper.makeId('led-backend-api'),
      description: 'APIs for LED app.',
      deployOptions: {
        stageName: helper.env
      }
    }
    if (helper.isDev) {
      // allow cors
      props = {
        ...props,
        defaultCorsPreflightOptions: {
          allowOrigins: apigw.Cors.ALL_ORIGINS,
          allowMethods: apigw.Cors.ALL_METHODS,
          statusCode: 200
        }
      }
    }
    const api = new apigw.RestApi(this, 'RestApi', props)
    const key = api.addApiKey('ApiKey')
    const plan = api.addUsagePlan('UsagePlan', {
      name: 'no-throttle',
      apiKey: key
    })
    plan.addApiStage({ stage: api.deploymentStage })
    return api
  }

  private addGetStatusMethod(
    resource: apigw.Resource,
    func: lambda.Function
  ): void {
    resource
      .addResource('status')
      .addMethod(
        'GET',
        new apigw.LambdaIntegration(func, this.integrationOptions()),
        this.methodOptions()
      )
  }

  private addSendMessageMethod(
    resource: apigw.Resource,
    func: lambda.Function
  ): void {
    resource
      .addResource('message')
      .addMethod(
        'POST',
        new apigw.LambdaIntegration(func, this.integrationOptions()),
        this.methodOptions()
      )
  }

  private integrationOptions(): apigw.LambdaIntegrationOptions {
    let options: apigw.LambdaIntegrationOptions = {
      passthroughBehavior: apigw.PassthroughBehavior.NEVER,
      proxy: false,
      integrationResponses: [
        {
          statusCode: '200'
        }
      ],
      requestTemplates: {
        'application/json':
          '{"message": $input.json("$"),"thingName":"$input.params(\'thing_name\')"}'
      }
    }
    if (helper.isDev) {
      // allow cors
      options = {
        ...options,
        integrationResponses: [
          {
            statusCode: '200',
            responseParameters: {
              'method.response.header.Access-Control-Allow-Headers':
                "'Content-Type,X-Api-Key'",
              'method.response.header.Access-Control-Allow-Origin': "'*'",
              'method.response.header.Access-Control-Allow-Methods':
                "'OPTIONS,GET,POST'"
            }
          }
        ]
      }
    }
    return options
  }

  private methodOptions(): apigw.MethodOptions {
    let options: apigw.MethodOptions = {
      apiKeyRequired: true,
      methodResponses: [
        {
          statusCode: '200'
        }
      ]
    }
    if (helper.isDev) {
      // allow cors
      options = {
        ...options,
        methodResponses: [
          {
            statusCode: '200',
            responseParameters: {
              'method.response.header.Access-Control-Allow-Headers': true,
              'method.response.header.Access-Control-Allow-Origin': true,
              'method.response.header.Access-Control-Allow-Methods': true
            }
          }
        ]
      }
    }
    return options
  }
}
