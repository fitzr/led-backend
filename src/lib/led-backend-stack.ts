import * as cdk from '@aws-cdk/core'
import { LedBackendApi } from './led-backend-api'
import { helper } from './stack-helper'
import { LedBackendLambda } from './led-backend-lambda'

export class LedBackendStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props)

    const lambda = new LedBackendLambda(this, helper.makeId('LedBackendLambda'))
    new LedBackendApi(this, helper.makeId('LedBackendApi'), lambda)
  }
}
