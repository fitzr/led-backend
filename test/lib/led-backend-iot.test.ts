import { SynthUtils } from '@aws-cdk/assert'
import { Stack } from '@aws-cdk/core'
import { LedBackendIot } from '../../src/lib/led-backend-iot'

test('LedBackendIot has no diff', () => {
  const stack = new Stack()
  new LedBackendIot(stack, 'LBI')
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
})
