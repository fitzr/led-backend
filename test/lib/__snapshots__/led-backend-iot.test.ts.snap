// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LedBackendIot has no diff 1`] = `
Object {
  "Resources": Object {
    "LBIIotPolicy46AE9730": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "iot:Connect",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:iot:undefined:",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":client/\${iot:Connection.Thing.ThingName}",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "iot:Publish",
                "iot:Receive",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:iot:undefined:",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":topic/$aws/things/\${iot:Connection.Thing.ThingName}/shadow/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "iot:Subscribe",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:iot:undefined:",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":topicfilter/$aws/things/\${iot:Connection.Thing.ThingName}/shadow/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PubSubShadowPolicy",
      },
      "Type": "AWS::IoT::Policy",
    },
  },
}
`;
